examples/vision/zero_dce.py

```diff
@@ -0,0 +1,444 @@
+        original_pool = self.pool(original_mean)
+        enhanced_pool = self.pool(enhanced_mean)
+
+        d_original_left = tf.nn.conv2d(
+            original_pool, self.left_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_original_right = tf.nn.conv2d(
+            original_pool, self.right_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_original_up = tf.nn.conv2d(
+            original_pool, self.up_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_original_down = tf.nn.conv2d(
+            original_pool, self.down_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+
+        d_enhanced_left = tf.nn.conv2d(
+            enhanced_pool, self.left_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_enhanced_right = tf.nn.conv2d(
+            enhanced_pool, self.right_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_enhanced_up = tf.nn.conv2d(
+            enhanced_pool, self.up_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+        d_enhanced_down = tf.nn.conv2d(
+            enhanced_pool, self.down_kernel, strides=[1, 1, 1, 1], padding="SAME"
+        )
+
+        d_left = tf.pow(d_original_left - d_enhanced_left, 2)
+        d_right = tf.pow(d_original_right - d_enhanced_right, 2)
+        d_up = tf.pow(d_original_up - d_enhanced_up, 2)
+        d_down = tf.pow(d_original_down - d_enhanced_down, 2)
+        return d_left + d_right + d_up + d_down
+
+
+"""
+### Deep Curve Estimation Model
+
+We implement the Zero-DCE framework as a Keras subclassed model.
+"""
+
+
+class ZeroDCE(keras.Model):
+    def __init__(self, **kwargs):
+        super(ZeroDCE, self).__init__(**kwargs)
+        self.dce_model = build_dce_net()
+
+    def compile(self, learning_rate, **kwargs):
+        super(ZeroDCE, self).compile(**kwargs)
+        self.optimizer = keras.optimizers.Adam(learning_rate=learning_rate)
+        self.spatial_constancy_loss = SpatialConsistancyLoss()
+
+    def get_enhanced_image(self, data, output):
+        r1 = output[:, :, :, :3]
+        r2 = output[:, :, :, 3:6]
+        r3 = output[:, :, :, 6:9]
+        r4 = output[:, :, :, 9:12]
+        r5 = output[:, :, :, 12:15]
+        r6 = output[:, :, :, 15:18]
+        r7 = output[:, :, :, 18:21]
+        r8 = output[:, :, :, 21:24]
+        x = data + r1 * (tf.pow(data, 2) - data)
+        x = x + r2 * (tf.pow(x, 2) - x)
+        x = x + r3 * (tf.pow(x, 2) - x)
+        enhanced_image = x + r4 * (tf.pow(x, 2) - x)
+        x = enhanced_image + r5 * (tf.pow(enhanced_image, 2) - enhanced_image)
+        x = x + r6 * (tf.pow(x, 2) - x)
+        x = x + r7 * (tf.pow(x, 2) - x)
+        enhanced_image = x + r8 * (tf.pow(x, 2) - x)
+        return enhanced_image
+
+    def call(self, data):
+        dce_net_output = self.dce_model(data)
+        return self.get_enhanced_image(data, dce_net_output)
+
+    def _calculate_losses(self, data, output):
```

Comment:
`compute_losses` would be more in line with other APIs. No need to make it private, either.

In fact this has no dependencies on `self` so it could be better as a standalone utility function rather than a method

Reply:
Renamed `_calculate_losses` to `compute_losses`. However, I didn't think that making it a standalone utility function would work since it has a dependency on `self.spatial_constancy_loss`. Instead, I made `get_enhanced_image` into a standalone utility function since it has not dependency on `self`.